### ISSUE TITLE
Handle namespaces in xml extractor [JIRA: RIAK-1705]

### ISSUE THREAD

**Issue by daurnimator** [basho/yokozuna#261]
At the moment elements with an xml namespace are silently discarded.

**Comment by rzezeski** [comment]
Yes, they are. I have a branch, somewhere, that fixes this. One question is what separator to use for namespaces. Should it be a `.` just like nested elements? Should it be `:` like in XML?

**Comment by daurnimator** [comment]
Well there are two different concepts: the namespace itself and it's alias ("qualified name").
It's probably best to drop the aliases (as chosen aliases may differ between documents) and only keep the namespaces.

In regards to choosing a seperator, I don't think using `.` will work well: too many clashes with children, however `:` doesn't work either: its the seperator for `field:query`
We should use a character not allowed in a namespace if possible?

**Comment by mattupstate** [comment]
For what its worth, the lxml library for Python uses curly braces for XML namespaces. For example:

``` python
xml.find("{http://mydomain.com/protocol-0}tagname")
```

**Comment by rzezeski** [comment]
@daurnimator: Given this example XML what would you expect the field
names to be? (I pulled this excerpt from section 6.2 of
http://www.w3.org/TR/2006/REC-xml-names11-20060816/)

``` XML
<?xml version="1.1"?>
<!-- initially, the default namespace is "books" -->
<book xmlns='urn:loc.gov:books'
      xmlns:isbn='urn:ISBN:0-395-36341-6'>
  <title>Cheaper by the Dozen</title>
  <isbn:number>1568491379</isbn:number>
  <notes>
    <!-- make HTML the default namespace for some commentary -->
    <p xmlns='http://www.w3.org/1999/xhtml'>
      This is a <i>funny</i> book!
    </p>
  </notes>
</book>
```

I modified the XML extractor to add the namespace (not the prefix, or
alias as you call it) to every field name by wrapping it in curly
braces and prepending it to the field name. Here's what the results
look like.

```
> Fields = yz_xml_extractor:extract(B).

> rp(Fields).
[{<<"{urn:loc.gov:books}book.{urn:loc.gov:books}notes.{http://www.w3.org/1999/xhtml}p">>,
  <<" book!\n    ">>},
 {<<"{urn:loc.gov:books}book.{urn:loc.gov:books}notes.{http://www.w3.org/1999/xhtml}p.{http://www.w3.org/1999/xhtml}i">>,
  <<"funny">>},
 {<<"{urn:loc.gov:books}book.{urn:loc.gov:books}notes.{http://www.w3.org/1999/xhtml}p">>,
  <<"\n      This is a ">>},
 {<<"{urn:loc.gov:books}book.{urn:ISBN:0-395-36341-6}number">>,
  <<"1568491379">>},
 {<<"{urn:loc.gov:books}book.{urn:loc.gov:books}title">>,
  <<"Cheaper by the Dozen">>}]
```

Is this really what you want? It seems like overkill to me. I feel
like using just the prefix would be enough. Do you have an example use
case with some real XML we can use as a strawman?

**Comment by daurnimator** [comment]
What you've got there seems okay... It is verbose, but at least it seems correct...

The namespace aliases shouldn't be used: often they are randomly generated by the xml serialiser: resulting in different documents and unusable search.

Here is a slightly modified example XMPP stanza from the MUC XEP (typical for our use-case of Riak)

``` xml
<message
    from='coven@chat.shakespeare.lit'
    id='367DCF6B-0CB4-482D-A142-C0B9E08016B5'
    to='wiccarocks@shakespeare.lit/laptop'
    type='groupchat'>
  <thread>e0ffe42b28561960c6b12b944a092794b9683a38</thread>
  <body>Thrice and once the hedge-pig whined.</body>
  <html xmlns='http://jabber.org/protocol/xhtml-im'>
    <body xmlns='http://www.w3.org/1999/xhtml'>
      <p style='font-weight:bold'>Thrice and once the hedge-pig whined.</p>
    </body>
  </html>
  <active xmlns='http://jabber.org/protocol/chatstates'/>
  <delay xmlns='urn:xmpp:delay'
     from='coven@chat.shakespeare.lit'
     stamp='2004-09-29T01:55:21Z'/>
</message>
```

**Comment by daurnimator** [comment]
Any progress on this? Sad not to see it in 0.14 :(

**Comment by rzezeski** [comment]
@daurnimator We are in a code freeze to finalize Riak 2.0. The focus is on bugs and other critical patches. This is not one of them for 2.0. The good news is I have a mostly working prototype and if you absolutely need this I can show you how to load and register the namespace aware extractor without requiring you to wait for the next release of Riak.

**Comment by daurnimator** [comment]
@rzezeski damn :(
If you could show off how to load the alternative extractor I'll give it a try.

**Comment by daurnimator** [comment]
Just found out the curly brace notation is know as "Clarks". More info here: http://www.jclark.com/xml/xmlns.htm

**Comment by DSomogyi** [comment]
Comment for Jira.
